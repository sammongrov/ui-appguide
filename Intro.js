import React from 'react';
import PropTypes from 'prop-types';

const Intro = ({ Layout, slides, onDone, onSkip }) => (
  <Layout slides={slides} onDone={() => onDone()} onSkip={() => onSkip()} />
);

Intro.propTypes = {
  Layout: PropTypes.func.isRequired,
  slides: PropTypes.arrayOf(PropTypes.object).isRequired,
  onDone: PropTypes.func.isRequired,
  onSkip: PropTypes.func.isRequired,
};

Intro.defaultProps = {};

export default Intro;
