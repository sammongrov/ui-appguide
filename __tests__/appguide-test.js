import React from 'react';
import renderer from 'react-test-renderer';
import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import { View, Text } from 'react-native';

import Intro from '../Intro';

const onDone = jest.fn();
const onSkip = jest.fn();

const Layout = ({ slides, onDone, onSkip }) => (
  <View>
    {slides.map((slide, index) => (
      <Text key={index} onPress={onSkip} onLongPress={onDone}>
        {slide.name}
      </Text>
    ))}
  </View>
);

const slides = [{ name: 'slide1' }, { name: 'slide2' }, { name: 'slide999' }];
const middlewares = [thunk];
const mockStore = configureStore(middlewares);

describe('appGuide renders correctly', () => {
  beforeEach(() => {
    onDone.mockClear();
  });

  it('with appGuide.viewed === false', () => {
    const initialState = {
      appGuide: {
        viewed: false,
      },
    };
    const store = mockStore(initialState);
    const tree = renderer
      .create(
        <Intro Layout={Layout} slides={slides} onDone={onDone} onSkip={onSkip} store={store} />,
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
    expect(onDone.mock.calls.length).toBe(0);
  });

  it('with appGuide.viewed === true', () => {
    const initialState = {
      appGuide: {
        viewed: true,
      },
    };
    const store = mockStore(initialState);
    const tree = renderer
      .create(
        <Intro Layout={Layout} slides={slides} onDone={onDone} onSkip={onSkip} store={store} />,
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
    expect(onDone.mock.calls.length).toBe(1);
    expect(onDone.mock.calls[0][0]).toBeUndefined();
  });

  it('with appGuide has no viewed property', () => {
    const initialState = {};
    const store = mockStore(initialState);
    const tree = renderer
      .create(
        <Intro Layout={Layout} slides={slides} onDone={onDone} onSkip={onSkip} store={store} />,
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
    expect(onDone.mock.calls.length).toBe(0);
  });
});

describe('appGuide dispatches onViewedIntro', () => {
  const initialState = {
    appGuide: {
      viewed: false,
    },
  };
  const store = mockStore(initialState);
  const component = renderer.create(
    <Intro Layout={Layout} slides={slides} onDone={onDone} onSkip={onSkip} store={store} />,
  ).root;

  beforeEach(() => {
    store.clearActions();
  });

  it('from onSkip prop', async () => {
    onSkip.mockClear();
    await component.findAllByType(Text)[0].props.onPress();

    expect(store.getActions()).toMatchSnapshot();
    expect(onSkip.mock.calls.length).toBe(1);
    expect(onSkip.mock.calls[0][0]).toBeUndefined();
  });

  it('from onDone prop', async () => {
    onDone.mockClear();
    await component.findAllByType(Text)[1].props.onLongPress();

    expect(store.getActions()).toMatchSnapshot();
    expect(onDone.mock.calls.length).toBe(1);
    expect(onDone.mock.calls[0][0]).toBeUndefined();
  });
});
