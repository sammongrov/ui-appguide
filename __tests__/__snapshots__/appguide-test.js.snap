// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`appGuide dispatches onViewedIntro from onDone prop 1`] = `Array []`;

exports[`appGuide dispatches onViewedIntro from onSkip prop 1`] = `Array []`;

exports[`appGuide renders correctly with appGuide has no viewed property 1`] = `
<View>
  <Text
    accessible={true}
    allowFontScaling={true}
    ellipsizeMode="tail"
    onLongPress={[Function]}
    onPress={[Function]}
  >
    slide1
  </Text>
  <Text
    accessible={true}
    allowFontScaling={true}
    ellipsizeMode="tail"
    onLongPress={[Function]}
    onPress={[Function]}
  >
    slide2
  </Text>
  <Text
    accessible={true}
    allowFontScaling={true}
    ellipsizeMode="tail"
    onLongPress={[Function]}
    onPress={[Function]}
  >
    slide999
  </Text>
</View>
`;

exports[`appGuide renders correctly with appGuide.viewed === false 1`] = `
<View>
  <Text
    accessible={true}
    allowFontScaling={true}
    ellipsizeMode="tail"
    onLongPress={[Function]}
    onPress={[Function]}
  >
    slide1
  </Text>
  <Text
    accessible={true}
    allowFontScaling={true}
    ellipsizeMode="tail"
    onLongPress={[Function]}
    onPress={[Function]}
  >
    slide2
  </Text>
  <Text
    accessible={true}
    allowFontScaling={true}
    ellipsizeMode="tail"
    onLongPress={[Function]}
    onPress={[Function]}
  >
    slide999
  </Text>
</View>
`;

exports[`appGuide renders correctly with appGuide.viewed === true 1`] = `
<View>
  <Text
    accessible={true}
    allowFontScaling={true}
    ellipsizeMode="tail"
    onLongPress={[Function]}
    onPress={[Function]}
  >
    slide1
  </Text>
  <Text
    accessible={true}
    allowFontScaling={true}
    ellipsizeMode="tail"
    onLongPress={[Function]}
    onPress={[Function]}
  >
    slide2
  </Text>
  <Text
    accessible={true}
    allowFontScaling={true}
    ellipsizeMode="tail"
    onLongPress={[Function]}
    onPress={[Function]}
  >
    slide999
  </Text>
</View>
`;
